File: lib/features/household/data/household_service.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:firebase_auth/firebase_auth.dart';
import 'package:flutter/foundation.dart';
import 'package:flutter_secure_storage/flutter_secure_storage.dart';
import 'package:uuid/uuid.dart';

import '../../../core/crypto/encryption_service.dart';
import '../../../core/crypto/key_manager.dart';

/// Service for household CRUD operations via Cloud Firestore.
///
/// Household names are end-to-end encrypted. Profile names in this service
/// are decrypted for display.
class HouseholdService {
  static final _db = FirebaseFirestore.instance;
  static final _auth = FirebaseAuth.instance;
  static const _uuid = Uuid();

  /// The [KeyManager] instance — must be set before calling household
  /// operations that need encryption/decryption.
  static KeyManager? keyManager;

  static String? get _uid => _auth.currentUser?.uid;

  /// The household key (if loaded).
  static String? get _key => keyManager?.householdKey;

  static String _enc(String plaintext) =>
      _key != null ? EncryptionService.encrypt(plaintext, _key!) : plaintext;

  // ignore: unused_element
  static String _dec(String ciphertext) =>
      _key != null ? EncryptionService.decrypt(ciphertext, _key!) : ciphertext;

  // ═══════════════════════════════════════════════════════════════════
  //  CREATE HOUSEHOLD
  // ═══════════════════════════════════════════════════════════════════

  /// Creates a new household and adds the current user as admin.
  ///
  /// Generates a per-household encryption key, encrypts the household name,
  /// and stores the key encrypted for the current user.
  ///
  /// Returns the created household data.
  static Future<Map<String, dynamic>> createHousehold(String name) async {
    final uid = _uid;
    if (uid == null) throw Exception('Not authenticated');

    final householdId = _uuid.v4();
    final now = DateTime.now();
    final km = keyManager ?? KeyManager.instance;

    // Generate a new household encryption key.
    final householdKey = await km.createHouseholdKey(householdId);

    // Now encrypt the household name with the new key.
    final encryptedName = EncryptionService.encrypt(name, householdKey);

    final batch = _db.batch();

    // Create household doc.
    batch.set(_db.collection('households').doc(householdId), {
      'id': householdId,
      'name': encryptedName,
      'created_by': uid,
      'created_at': now.toIso8601String(),
    });

    // Add creator as admin member.
    // Doc ID = {uid}_{householdId} — deterministic for Firestore security rules.
    final memberId = '${uid}_$householdId';
    batch.set(_db.collection('household_members').doc(memberId), {
      'household_id': householdId,
      'user_id': uid,
      'role': 'admin',
      'joined_at': now.toIso8601String(),
    });

    await batch.commit();

    debugPrint('[HouseholdService] ✓ Created household $householdId');

    // Encrypt and store the profile name now that we have a household key.
    await _encryptProfileName(uid, householdKey);

    return {
      'id': householdId,
      'name': name,
      'created_by': uid,
      'created_at': now.toIso8601String(),
    };
  }

  // ═══════════════════════════════════════════════════════════════════
  //  GET CURRENT HOUSEHOLD
  // ═══════════════════════════════════════════════════════════════════

  /// Fetches the current user's household (if any).
  ///
  /// Also loads the household encryption key so data can be decrypted.
  /// Returns null if the user hasn't joined a household yet.
  static Future<Map<String, dynamic>?> getCurrentHousehold() async {
    final uid = _uid;
    if (uid == null) return null;

    try {
      // Find the user's household membership.
      final memberSnap = await _db
          .collection('household_members')
          .where('user_id', isEqualTo: uid)
          .limit(1)
          .get();

      if (memberSnap.docs.isEmpty) return null;

      final membership = memberSnap.docs.first.data();
      final householdId = membership['household_id'] as String;

      // Load the household encryption key.
      final km = keyManager ?? KeyManager.instance;
      await km.loadHouseholdKey(householdId);

      // Fetch the household doc.
      final householdDoc =
          await _db.collection('households').doc(householdId).get();
      if (!householdDoc.exists) return null;

      final householdData = householdDoc.data()!;

      // Decrypt name using the now-loaded key.
      final decryptedName = km.householdKey != null
          ? EncryptionService.decrypt(
              householdData['name'] as String, km.householdKey!)
          : householdData['name'] as String;

      return {
        'membership': membership,
        'household': {
          'id': householdId,
          'name': decryptedName,
          'created_by': householdData['created_by'],
          'created_at': householdData['created_at'],
        },
        'role': membership['role'],
      };
    } catch (e) {
      debugPrint('[HouseholdService] ✗ getCurrentHousehold: $e');
      return null;
    }
  }

  // ═══════════════════════════════════════════════════════════════════
  //  MEMBERS
  // ═══════════════════════════════════════════════════════════════════

  /// Fetches all members of a household with their profile info.
  static Future<List<Map<String, dynamic>>> getHouseholdMembers(
      String householdId) async {
    // Ensure the household key is loaded so profile names can be decrypted.
    final km = keyManager ?? KeyManager.instance;
    await km.loadHouseholdKey(householdId);

    final memberSnap = await _db
        .collection('household_members')
        .where('household_id', isEqualTo: householdId)
        .get();

    final results = <Map<String, dynamic>>[];

    for (final doc in memberSnap.docs) {
      final data = doc.data();
      final userId = data['user_id'] as String;

      // Fetch profile.
      Map<String, dynamic>? profile;
      final profileDoc = await _db.collection('profiles').doc(userId).get();
      if (profileDoc.exists) {
        final pData = profileDoc.data()!;
        final rawName = pData['full_name'] as String?;
        // Decrypt using the key we just loaded — don't rely on the static
        // _key getter which may be null if keyManager was not set.
        profile = {
          'id': userId,
          'full_name': (km.householdKey != null && rawName != null && rawName.isNotEmpty)
              ? EncryptionService.decryptNullable(rawName, km.householdKey!)
              : rawName,
          'avatar_url': pData['avatar_url'] as String?,
        };
      }

      results.add({
        'user_id': userId,
        'role': data['role'],
        'joined_at': data['joined_at'],
        'profiles': profile,
      });
    }

    return results;
  }

  // ═══════════════════════════════════════════════════════════════════
  //  INVITES
  // ═══════════════════════════════════════════════════════════════════

  /// Invites a user to the household by email.
  static Future<void> inviteByEmail({
    required String householdId,
    required String email,
  }) async {
    final id = _uuid.v4();
    await _db.collection('household_invites').doc(id).set({
      'id': id,
      'household_id': householdId,
      'invited_email': email,
      'invited_by': _uid,
      'status': 'pending',
      'created_at': DateTime.now().toIso8601String(),
    });
  }

  /// Checks if the current user has a pending invite and accepts it.
  ///
  /// When accepting, the inviter's device must share the household key
  /// with the new member. For now, we handle this by having the accepting
  /// user request the key from an existing member's device.
  static Future<Map<String, dynamic>?> checkAndAcceptInvite() async {
    final user = _auth.currentUser;
    if (user == null || user.email == null) return null;

    final inviteSnap = await _db
        .collection('household_invites')
        .where('invited_email', isEqualTo: user.email!)
        .where('status', isEqualTo: 'pending')
        .limit(1)
        .get();

    if (inviteSnap.docs.isEmpty) return null;

    final invite = inviteSnap.docs.first.data();
    final householdId = invite['household_id'] as String;

    final batch = _db.batch();

    // Add user as member.
    // Doc ID = {uid}_{householdId} — deterministic for Firestore security rules.
    final memberId = '${user.uid}_$householdId';
    batch.set(_db.collection('household_members').doc(memberId), {
      'household_id': householdId,
      'user_id': user.uid,
      'role': 'member',
      'joined_at': DateTime.now().toIso8601String(),
    });

    // Mark invite as accepted.
    batch.update(inviteSnap.docs.first.reference, {'status': 'accepted'});

    await batch.commit();

    // Load the household key and encrypt the profile name.
    final km = keyManager ?? KeyManager.instance;
    final hKey = await km.loadHouseholdKey(householdId);
    if (hKey != null) {
      await _encryptProfileName(user.uid, hKey);
    }

    // Fetch the household info.
    final householdDoc =
        await _db.collection('households').doc(householdId).get();
    if (!householdDoc.exists) return null;

    final hData = householdDoc.data()!;
    return {
      'id': householdId,
      'name': hData['name'], // Still encrypted — caller will decrypt.
    };
  }

  /// Removes a member from the household.
  static Future<void> removeMember({
    required String householdId,
    required String userId,
  }) async {
    final batch = _db.batch();

    // Direct delete using deterministic doc ID.
    batch.delete(
      _db.collection('household_members').doc('${userId}_$householdId'),
    );

    // Also clean up any legacy random-UUID member docs (migration safety).
    final legacySnap = await _db
        .collection('household_members')
        .where('household_id', isEqualTo: householdId)
        .where('user_id', isEqualTo: userId)
        .get();
    for (final doc in legacySnap.docs) {
      if (doc.id != '${userId}_$householdId') {
        batch.delete(doc.reference);
      }
    }

    // Also delete the member's household key.
    final keySnap = await _db
        .collection('household_keys')
        .where('household_id', isEqualTo: householdId)
        .where('user_id', isEqualTo: userId)
        .get();
    for (final doc in keySnap.docs) {
      batch.delete(doc.reference);
    }

    await batch.commit();
  }

  /// Reads the locally-cached profile name, encrypts it with the household
  /// key, and writes it to the Firestore profile doc.
  static Future<void> _encryptProfileName(String uid, String householdKey) async {
    try {
      const secureStorage = FlutterSecureStorage();
      final localName = await secureStorage.read(key: 'profile_name_$uid');
      if (localName != null && localName.isNotEmpty) {
        final encryptedName = EncryptionService.encrypt(localName, householdKey);
        await _db.collection('profiles').doc(uid).update({
          'full_name': encryptedName,
        });
        debugPrint('[HouseholdService] ✓ Encrypted profile name for $uid');
      }
    } catch (e) {
      debugPrint('[HouseholdService] ✗ Failed to encrypt profile name: $e');
    }
  }

  /// Updates the household name (encrypted).
  static Future<void> updateHouseholdName({
    required String householdId,
    required String name,
  }) async {
    await _db.collection('households').doc(householdId).update({
      'name': _enc(name),
    });
  }

  // ═══════════════════════════════════════════════════════════════════
  //  DATA MIGRATION
  // ═══════════════════════════════════════════════════════════════════

  /// Migrates household_members docs from random UUIDs to deterministic
  /// `{userId}_{householdId}` doc IDs.
  ///
  /// This is required for Firestore security rules that use `exists()` to
  /// verify household membership. Safe to call multiple times (idempotent).
  static Future<void> migrateMemberDocIds() async {
    final uid = _uid;
    if (uid == null) return;

    try {
      // Find all memberships for the current user.
      final snap = await _db
          .collection('household_members')
          .where('user_id', isEqualTo: uid)
          .get();

      for (final doc in snap.docs) {
        final data = doc.data();
        final householdId = data['household_id'] as String;
        final expectedId = '${uid}_$householdId';

        // Skip if already using the correct deterministic ID.
        if (doc.id == expectedId) continue;

        debugPrint(
          '[HouseholdService] Migrating member doc ${doc.id} → $expectedId',
        );

        final batch = _db.batch();

        // Create new doc with deterministic ID.
        batch.set(
          _db.collection('household_members').doc(expectedId),
          data,
        );

        // Delete the old random-UUID doc.
        batch.delete(doc.reference);

        await batch.commit();
      }
    } catch (e) {
      debugPrint('[HouseholdService] ✗ migrateMemberDocIds: $e');
    }
  }
}

File: PacelliApp/Sources/Core/HouseholdService.swift
import FirebaseAuth
import FirebaseFirestore
import Foundation
import PacelliKit

struct CurrentHousehold: Sendable {
    let household: Household
    let role: String
}

/// Household CRUD. Port of `lib/features/household/data/household_service.dart`
/// — identical collections, doc IDs, field names, and encryption points.
enum HouseholdService {
    private static var db: Firestore { Firestore.firestore() }
    private static var uid: String? { Auth.auth().currentUser?.uid }

    /// Creates a household and adds the current user as admin.
    /// Batch parity with Dart: `households/{uuid}` (encrypted name, ISO date)
    /// + `household_members/{uid}_{householdId}` (deterministic ID — rules).
    static func createHousehold(named name: String) async throws -> CurrentHousehold {
        guard let uid else { throw PacelliError.notSignedIn }

        let householdId = UUID().uuidString.lowercased()
        let now = Date()

        let householdKey = try await KeyManager.shared.createHouseholdKey(householdId)
        let encryptedName = try PacelliCrypto.encrypt(name, key: householdKey)

        let batch = db.batch()
        batch.setData(
            [
                "id": householdId,
                "name": encryptedName,
                "created_by": uid,
                "created_at": DartISO8601.string(from: now),
            ],
            forDocument: db.collection("households").document(householdId))

        let member = HouseholdMember(
            userId: uid, householdId: householdId, role: "admin", joinedAt: now)
        batch.setData(
            member.toMap(),
            forDocument: db.collection("household_members").document(member.documentID))

        try await batch.commit()

        // Dart parity (_encryptProfileName): now that a household key exists,
        // encrypt the locally-cached display name into the profile doc.
        await encryptProfileName(uid: uid, householdKey: householdKey)

        return CurrentHousehold(
            household: Household(id: householdId, name: name, createdBy: uid, createdAt: now),
            role: "admin")
    }


    /// Content collections that make a household non-empty. Same list as
    /// `BurnService.wipeHousehold` — if a household-scoped collection is added,
    /// it must be added in both places or an "empty" household could still
    /// hold data when it is discarded.
    static let householdContentCollections = [
        "tasks", "checklists", "scratch_plans",
        "task_categories", "task_attachments", "plan_attachments",
        "inventory_items", "inventory_categories",
        "inventory_locations", "inventory_logs", "inventory_attachments",
        "manual_entries", "manual_categories", "feedback", "diagnostics",
        "weekly_digests",
        "subtasks", "checklist_items", "plan_entries", "plan_checklist_items",
    ]

    /// Deletes the caller's own auto-provisioned households that are provably
    /// empty, keeping `keep`. Returns how many were discarded.
    ///
    /// Someone who used the app before being invited is left holding a second,
    /// empty household forever; they accumulate, and every extra membership is
    /// another way for `getCurrentHousehold` to pick wrong. Deliberately
    /// strict: it discards only a household the caller created, with no other
    /// member, and zero documents in EVERY content collection. Any doubt and
    /// it leaves the household alone. Non-fatal throughout — failing to tidy
    /// up must never break a successful join.
    @discardableResult
    static func discardOwnEmptyHouseholds(except keep: String) async -> Int {
        guard let uid else { return 0 }
        var discarded = 0
        do {
            let memberships = try await db.collection("household_members")
                .whereField("user_id", isEqualTo: uid)
                .getDocuments()
            for memberDoc in memberships.documents {
                guard let hid = memberDoc.data()["household_id"] as? String,
                      hid != keep, !hid.isEmpty
                else { continue }
                if await isDiscardable(hid, uid: uid) {
                    await discard(hid, uid: uid)
                    discarded += 1
                }
            }
        } catch {
            print("[HouseholdService] discardOwnEmptyHouseholds failed: \(error)")
        }
        return discarded
    }

    private static func isDiscardable(_ householdId: String, uid: String) async -> Bool {
        do {
            let household = try await db.collection("households")
                .document(householdId).getDocument()
            // Never touch a household someone else created.
            guard household.exists,
                  household.data()?["created_by"] as? String == uid
            else { return false }

            let members = try await db.collection("household_members")
                .whereField("household_id", isEqualTo: householdId)
                .getDocuments()
            guard members.documents.count == 1 else { return false }

            // A pending invite means somebody is expected — not abandoned.
            let invites = try await db.collection("household_invites")
                .whereField("household_id", isEqualTo: householdId)
                .limit(to: 1)
                .getDocuments()
            guard invites.documents.isEmpty else { return false }

            for collection in householdContentCollections {
                let snap = try await db.collection(collection)
                    .whereField("household_id", isEqualTo: householdId)
                    .limit(to: 1)
                    .getDocuments()
                guard snap.documents.isEmpty else { return false }
            }
            return true
        } catch {
            print("[HouseholdService] isDiscardable(\(householdId)) failed: \(error)")
            return false
        }
    }

    private static func discard(_ householdId: String, uid: String) async {
        do {
            // Join codes first: they are the only thing that could let someone
            // else walk into a household we are about to delete.
            try? await JoinCodeService.revokeAll(householdId: householdId)
            await KeyManager.shared.deleteKeyFromFirestore(householdId)
            let batch = db.batch()
            batch.deleteDocument(
                db.collection("household_members").document("\(uid)_\(householdId)"))
            batch.deleteDocument(db.collection("households").document(householdId))
            try await batch.commit()
            SecureStore.delete("hk_\(householdId)")
            print("[HouseholdService] discarded empty household \(householdId)")
        } catch {
            print("[HouseholdService] discard(\(householdId)) failed: \(error)")
        }
    }

    /// Reads the Keychain-cached profile name, encrypts it with the
    /// household key, writes it to `profiles/{uid}`. Non-fatal on failure.
    ///
    /// Was `private` and called only from `createHousehold`, which meant a
    /// name could only ever reach Firestore for the person who FOUNDED a
    /// household. Anyone who joined by code or invite stayed nameless forever
    /// and showed up in the members list as "Member". Now also called on join
    /// and whenever a session lands in a household.
    static func encryptProfileName(uid: String, householdKey: String) async {
        guard let localName = SecureStore.read("profile_name_\(uid)"),
              !localName.isEmpty,
              let encrypted = try? PacelliCrypto.encrypt(localName, key: householdKey)
        else { return }
        try? await db.collection("profiles").document(uid)
            .updateData(["full_name": encrypted])
    }

    /// Set (or change) the name other members see.
    ///
    /// Sign in with Apple hands back a name only on the very FIRST
    /// authorization, ever — sign in a second time and Apple returns nothing,
    /// permanently, unless the user revokes the app in iOS Settings. Relying
    /// on it meant most people could never have a name at all. This is the
    /// path that always works.
    ///
    /// Encrypted with the household key like every other piece of content, so
    /// the server never sees it. Cached in the Keychain too, so it survives
    /// into any household joined later.
    static func setDisplayName(_ name: String, householdId: String) async throws {
        guard let uid = Auth.auth().currentUser?.uid else { throw PacelliError.notSignedIn }
        let trimmed = name.trimmingCharacters(in: .whitespacesAndNewlines)
        guard let key = await KeyManager.shared.loadHouseholdKey(householdId) else {
            throw PacelliError.missingHouseholdKey
        }

        if trimmed.isEmpty {
            SecureStore.delete("profile_name_\(uid)")
            try await db.collection("profiles").document(uid)
                .setData(["full_name": ""], merge: true)
            return
        }
        SecureStore.write("profile_name_\(uid)", value: trimmed)
        let encrypted = try PacelliCrypto.encrypt(trimmed, key: key)
        // merge:true, not updateData — a profile doc can be missing for an
        // account created before profiles existed, and updateData would throw.
        try await db.collection("profiles").document(uid)
            .setData(["full_name": encrypted], merge: true)
    }

    /// The current user's own name, decrypted, for pre-filling the field.
    static func currentDisplayName(householdId: String) async -> String {
        guard let uid = Auth.auth().currentUser?.uid else { return "" }
        if let cached = SecureStore.read("profile_name_\(uid)"), !cached.isEmpty {
            return cached
        }
        guard let key = await KeyManager.shared.loadHouseholdKey(householdId),
              let data = try? await db.collection("profiles").document(uid).getDocument().data(),
              let encrypted = data["full_name"] as? String, !encrypted.isEmpty
        else { return "" }
        return (try? PacelliCrypto.decrypt(encrypted, key: key)) ?? ""
    }

    /// Renames the household (any member — flat-membership model; Firestore
    /// rules allow `update` for `isMember`). Name is E2E-encrypted with the
    /// household key before upload, same as `createHousehold`.
    static func renameHousehold(_ householdId: String, to name: String) async throws {
        let trimmed = name.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return }
        guard let key = await KeyManager.shared.loadHouseholdKey(householdId) else {
            throw PacelliError.missingHouseholdKey
        }
        let encrypted = try PacelliCrypto.encrypt(trimmed, key: key)
        try await db.collection("households").document(householdId)
            .updateData(["name": encrypted])
    }

    /// The current user's household with its key loaded, or nil if none.
    /// - Parameter preferring: which household to land in when the user
    ///   belongs to more than one. Someone who used guest mode before being
    ///   invited ends up with two member docs; the old `limit(to: 1)` made the
    ///   winner whichever doc Firestore returned first, so a freshly accepted
    ///   invite could silently drop them back into their own empty household.
    ///   Without a preference we now take the most recently joined household,
    ///   which is at least deterministic.
    static func getCurrentHousehold(preferring preferredId: String? = nil) async
        -> CurrentHousehold?
    {
        guard let uid else { return nil }
        do {
            print("[HouseholdService] membership query for \(uid)…")
            let memberSnap = try await db.collection("household_members")
                .whereField("user_id", isEqualTo: uid)
                .getDocuments()
            print("[HouseholdService] membership docs: \(memberSnap.documents.count)")

            let memberships = memberSnap.documents.map { $0.data() }
            let preferred = preferredId.flatMap { wanted in
                memberships.first { $0["household_id"] as? String == wanted }
            }
            let newest = memberships.max { a, b in
                let lhs = DartISO8601.date(from: a["joined_at"] as? String) ?? .distantPast
                let rhs = DartISO8601.date(from: b["joined_at"] as? String) ?? .distantPast
                return lhs < rhs
            }
            guard let membership = preferred ?? newest,
                  let householdId = membership["household_id"] as? String
            else { return nil }

            let key = await KeyManager.shared.loadHouseholdKey(householdId)

            let doc = try await db.collection("households").document(householdId).getDocument()
            guard var data = doc.data() else { return nil }

            if let key, let encryptedName = data["name"] as? String {
                data["name"] = PacelliCrypto.decryptNullable(encryptedName, key: key)
                    ?? encryptedName
            }
            data["id"] = householdId

            guard let household = Household(map: data) else { return nil }
            return CurrentHousehold(
                household: household,
                role: membership["role"] as? String ?? "member")
        } catch {
            print("[HouseholdService] getCurrentHousehold failed: \(error)")
            return nil
        }
    }
}
